File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Doms/Protocol/DomsTotalsHandler.cs
using FccDesktopAgent.Core.Adapter.Common;
using FccDesktopAgent.Core.Adapter.Doms.Jpl;
using FccDesktopAgent.Core.Adapter.Doms.Mapping;

namespace FccDesktopAgent.Core.Adapter.Doms.Protocol;

/// <summary>
/// JPL message builder for pump totals queries.
/// Ported from legacy FpTotals (EXTC 0x09): FpId, TotalVol, TotalMoney.
/// </summary>
internal static class DomsTotalsHandler
{
    public const string TotalsRequest = "FpTotals_req";
    public const string TotalsResponse = "FpTotals_resp";

    /// <summary>Build FpTotals_req. fpId=0 means all pumps.</summary>
    public static JplMessage BuildTotalsRequest(int fpId = 0)
        => new(Name: TotalsRequest, Data: new Dictionary<string, string>
        {
            ["FpId"] = fpId.ToString(),
        });

    /// <summary>
    /// Parse FpTotals_resp into PumpTotals records.
    /// Data contains: FpId, TotalVol (centilitres), TotalMoney (DOMS x10).
    /// </summary>
    public static IReadOnlyList<PumpTotals> ParseTotalsResponse(
        JplMessage response,
        string currencyCode,
        int pumpNumberOffset)
    {
        if (response.Name != TotalsResponse) return [];

        var data = response.Data;
        if (data is null) return [];

        if (!data.TryGetValue("FpId", out var fpIdStr) || !int.TryParse(fpIdStr, out var fpId))
            return [];

        var totalVolCl = data.TryGetValue("TotalVol", out var volStr) && long.TryParse(volStr, out var vol)
            ? vol : 0L;
        var totalMoneyX10 = data.TryGetValue("TotalMoney", out var monStr) && long.TryParse(monStr, out var mon)
            ? mon : 0L;

        return
        [
            new PumpTotals(
                PumpNumber: fpId + pumpNumberOffset,
                TotalVolumeMicrolitres: DomsCanonicalMapper.CentilitresToMicrolitres(totalVolCl),
                TotalAmountMinorUnits: DomsCanonicalMapper.DomsAmountToMinorUnits(totalMoneyX10),
                CurrencyCode: currencyCode,
                ObservedAtUtc: DateTimeOffset.UtcNow)
        ];
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/doms/protocol/DomsTotalsHandler.kt
package com.fccmiddleware.edge.adapter.doms.protocol

import com.fccmiddleware.edge.adapter.common.PumpTotals
import com.fccmiddleware.edge.adapter.doms.jpl.JplMessage
import com.fccmiddleware.edge.adapter.doms.mapping.DomsCanonicalMapper
import java.time.Instant

/**
 * JPL message builder for pump totals queries.
 * Ported from legacy FpTotals (EXTC 0x09): FpId, TotalVol, TotalMoney.
 */
object DomsTotalsHandler {
    const val TOTALS_REQUEST = "FpTotals_req"
    const val TOTALS_RESPONSE = "FpTotals_resp"

    fun buildTotalsRequest(fpId: Int = 0): JplMessage =
        JplMessage(name = TOTALS_REQUEST, data = mapOf("FpId" to fpId.toString()))

    fun parseTotalsResponse(
        response: JplMessage,
        currencyCode: String,
        pumpNumberOffset: Int,
    ): List<PumpTotals> {
        if (response.name != TOTALS_RESPONSE) return emptyList()
        val data = response.data

        val fpId = data["FpId"]?.toIntOrNull() ?: return emptyList()
        val totalVolCl = data["TotalVol"]?.toLongOrNull() ?: 0L
        val totalMoneyX10 = data["TotalMoney"]?.toLongOrNull() ?: 0L

        return listOf(
            PumpTotals(
                pumpNumber = fpId + pumpNumberOffset,
                totalVolumeMicrolitres = DomsCanonicalMapper.centilitresToMicrolitres(totalVolCl),
                totalAmountMinorUnits = DomsCanonicalMapper.domsAmountToMinorUnits(totalMoneyX10),
                currencyCode = currencyCode,
                observedAtUtc = Instant.now().toString(),
            )
        )
    }
}
